server/middleware/auth/authentication.go

```diff
@@ -0,0 +1,226 @@
+	url     map[string][]string
+}
+
+func NewMiddleware(svc *service.Service) *AuthenticationMiddleware {
+	return &AuthenticationMiddleware{
+		service: svc,
+		url:     make(map[string][]string),
+	}
+}
+
+func (amw *AuthenticationMiddleware) AddURL(path string, methods ...string) {
+	// Router is only used to convert path to regex.
+	pathRegex, err := mux.NewRouter().Path(path).Methods(http.MethodGet).GetPathRegexp()
+	if err != nil {
+		panic(err)
+	}
+	m, ok := amw.url[pathRegex]
+	if !ok {
+		amw.url[pathRegex] = methods
+	} else {
+		amw.url[pathRegex] = append(m, methods...)
+	}
+}
+
+// Middleware supports signature authentication and responds to the client by generating information necessary for authentication.
+func (amw *AuthenticationMiddleware) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// if it is not an authentication URL, the authentication check is not performed.
+		if !amw.isAuthenticationURL(r) {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		sigAuthParts, err := ParseSignatureAuthorizationParts(r.Header.Get("Authorization"))
+		//signatureAuthentication, err := parseSignatureAuthentication(r)
+		if err != nil {
+			// result code to 400. Bad request
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		err = basicValidate(sigAuthParts)
+		if err != nil {
+			// result code to 400. Bad request
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if sigAuthParts[Nonce] == "" {
+			err = amw.generateAuthenticationAndSetHeader(w, sigAuthParts)
+			if err != nil {
+				log.Error("failed to generate authentication", err)
+				http.Error(w, "failed to generate authentication", http.StatusInternalServerError)
+			}
+			http.Error(w, "", http.StatusUnauthorized)
+			return
+		}
+
+		cache := amw.service.Cache
+		auth := cache.Get(sigAuthParts[KeyId], sigAuthParts[Nonce])
+		if auth == nil {
+			// expired
+			err = amw.generateAuthenticationAndSetHeader(w, sigAuthParts)
+			if err != nil {
+				log.Error("failed to generate authentication", err)
+				http.Error(w, "failed to generate authentication", http.StatusInternalServerError)
+			}
+			http.Error(w, "", http.StatusUnauthorized)
+			return
+		}
+
+		nonce := sigAuthParts[Nonce]
+		signature, err := base64.StdEncoding.DecodeString(sigAuthParts[Signature])
+		if err != nil {
+			http.Error(w, "failed to decode signature", http.StatusBadRequest)
+			return
+		}
+
+		requesterAddress := sigAuthParts[KeyId]
+		pubKey, err := amw.service.PanaceaClient.GetPubKey(requesterAddress)
+		if err != nil {
+			log.Error(fmt.Sprintf("failed to get the account's public key account: %s", requesterAddress), err)
+			http.Error(w, "failed to get the account's public key", http.StatusInternalServerError)
+			return
+		}
+
+		if !pubKey.VerifySignature([]byte(nonce), signature) {
+			http.Error(w, "failed to verification signature", http.StatusBadRequest)
+			return
+		}
+
```

Comment:
After verifying the signature successfully, the nonce should be removed from cache for preventing replay attack.

Reply:
Your right. I will fix it.